vslib/src/SwitchStateBaseMACsec.cpp

```diff
@@ -0,0 +1,831 @@
+    {
+        if (m_macsecManager.delete_macsec_sa(macsecAttr))
+        {
+            SWSS_LOG_NOTICE(
+                "The MACsec SA %s:%u at the device %s is deleted.",
+                macsecAttr.m_sci.c_str(),
+                static_cast<std::uint32_t>(macsecAttr.m_an),
+                macsecAttr.m_macsecName.c_str());
+        }
+    }
+
+    auto sid = sai_serialize_object_id(macsecSaId);
+    return remove_internal(SAI_OBJECT_TYPE_MACSEC_SA, sid);
+}
+
+sai_status_t SwitchStateBase::getACLTable(
+    _In_ sai_object_id_t entryId,
+    _Out_ sai_object_id_t &tableId)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    // Find ACL Table
+    attr.id = SAI_ACL_ENTRY_ATTR_TABLE_ID;
+    attr.value.oid = SAI_NULL_OBJECT_ID;
+
+    CHECK_STATUS(get(SAI_OBJECT_TYPE_ACL_ENTRY,entryId,1,&attr));
+
+    tableId = attr.value.oid;
+
+    return SAI_STATUS_SUCCESS;
+}
+
+sai_status_t SwitchStateBase::findPortByMACsecFlow(
+        _In_ sai_object_id_t macsecFlowId,
+        _Out_ sai_object_id_t &portId)
+{
+    SWSS_LOG_ENTER();
+
+    std::vector<sai_attribute_t> attrs;
+    sai_attribute_t attr;
+
+    // MACsec flow => ACL entry => ACL table
+    // MACsec flow => MACsec SC
+    // (ACL table & MACsec SC) => port
+
+    // Find ACL Entry
+    attr.id = SAI_ACL_ENTRY_ATTR_ACTION_MACSEC_FLOW;
+    attr.value.aclaction.parameter.oid = macsecFlowId;
+    attr.value.aclaction.enable = true;
+    std::vector<sai_object_id_t> aclEntryIds;
+    findObjects(SAI_OBJECT_TYPE_ACL_ENTRY, attr, aclEntryIds);
+
+    if (aclEntryIds.empty())
+    {
+        // No ACL Entry is bound to the flow
+        SWSS_LOG_DEBUG(
+            "Cannot find corresponding ACL entry for the flow %s",
+            sai_serialize_object_id(macsecFlowId).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    // Find ACL Table
+    sai_object_id_t aclTableId;
+
+    // All ACL entries with same MACsec flow correspond one ACL Table
+    // Because an ACL Table corresponds one MACsec Port
+    // And a flow never belongs to two MACsec Port
+    if (getACLTable(aclEntryIds.front(), aclTableId) != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR(
+            "Cannot find corresponding ACL table for the entry %s",
+            sai_serialize_object_id(aclEntryIds.front()).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    // Find MACsec SC
+    attr.id = SAI_MACSEC_SC_ATTR_FLOW_ID;
+    attr.value.oid = macsecFlowId;
+    std::vector<sai_object_id_t> macsecScIds;
+    findObjects(SAI_OBJECT_TYPE_MACSEC_SC, attr, macsecScIds);
+
+    // One MACsec SC will only belong to one MACsec flow
+    // Meanwhile one MACsec flow will just belong to one port
+    if (macsecScIds.empty())
+    {
+        SWSS_LOG_ERROR(
+            "Cannot find corresponding MACsec SC for the flow %s",
+            sai_serialize_object_id(macsecFlowId).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    attrs.resize(1);
```

Comment:
declare std::vector<sai_attribute_t> attrs; as close as possible to first use

Reply:
and you dont need this vector at all, sincie you are only using 1 attribute any time, single "attr" that you have declared above is good enough